partiql-lang/src/main/kotlin/org/partiql/lang/planner/transforms/plan/RexConverter.kt

```diff
@@ -262,59 +265,118 @@ internal object RexConverter : PartiqlAst.VisitorFold<RexConverter.Ctx>() {
+            rhs = rhs,
             op = Rex.Binary.Op.EQ,
             type = StaticType.BOOL,
         )
     }
 
     override fun walkExprNe(node: PartiqlAst.Expr.Ne, ctx: Ctx) = visit(node) {
+        val (lhs, rhs) = walkComparisonOperands(node.operands)
         Plan.rexBinary(
-            lhs = convert(node.operands[0]),
-            rhs = convert(node.operands[1]),
+            lhs = lhs,
+            rhs = rhs,
             op = Rex.Binary.Op.NEQ,
             type = StaticType.BOOL,
         )
     }
 
     override fun walkExprGt(node: PartiqlAst.Expr.Gt, ctx: Ctx) = visit(node) {
+        val (lhs, rhs) = walkComparisonOperands(node.operands)
         Plan.rexBinary(
-            lhs = convert(node.operands[0]),
-            rhs = convert(node.operands[1]),
+            lhs = lhs,
+            rhs = rhs,
             op = Rex.Binary.Op.GT,
             type = StaticType.BOOL,
         )
     }
 
     override fun walkExprGte(node: PartiqlAst.Expr.Gte, ctx: Ctx) = visit(node) {
+        val (lhs, rhs) = walkComparisonOperands(node.operands)
         Plan.rexBinary(
-            lhs = convert(node.operands[0]),
-            rhs = convert(node.operands[1]),
+            lhs = lhs,
+            rhs = rhs,
             op = Rex.Binary.Op.GTE,
             type = StaticType.BOOL,
         )
     }
 
     override fun walkExprLt(node: PartiqlAst.Expr.Lt, ctx: Ctx) = visit(node) {
+        val (lhs, rhs) = walkComparisonOperands(node.operands)
         Plan.rexBinary(
-            lhs = convert(node.operands[0]),
-            rhs = convert(node.operands[1]),
+            lhs = lhs,
+            rhs = rhs,
             op = Rex.Binary.Op.LT,
             type = StaticType.BOOL,
         )
     }
 
     override fun walkExprLte(node: PartiqlAst.Expr.Lte, ctx: Ctx) = visit(node) {
+        val (lhs, rhs) = walkComparisonOperands(node.operands)
         Plan.rexBinary(
-            lhs = convert(node.operands[0]),
-            rhs = convert(node.operands[1]),
+            lhs = lhs,
+            rhs = rhs,
             op = Rex.Binary.Op.LTE,
             type = StaticType.BOOL,
         )
     }
 
+    /**
+     * Converts Comparison Operands. Also coerces them if one of them is an array.
+     */
+    private fun walkComparisonOperands(operands: List<PartiqlAst.Expr>): Pair<Rex, Rex> {
+        var lhs = convert(operands[0])
+        var rhs = convert(operands[1])
+        if (lhs is Rex.Collection.Array) { rhs = coercePotentialSubquery(rhs) }
+        if (rhs is Rex.Collection.Array) { lhs = coercePotentialSubquery(lhs) }
+        return lhs to rhs
+    }
+
+    /**
+     * We convert the scalar subquery of a SFW into a scalar subquery of a SELECT VALUE.
+     */
+    private fun coercePotentialSubquery(rex: Rex): Rex {
+        var rhs = rex
+        if (rhs is Rex.Query.Scalar.Subquery) {
+            val sfw = rhs.query as? Rex.Query.Collection ?: error("Malformed plan, all scalar subqueries should hold collection queries")
```

Comment:
Does it matter which `Rex.Query.Collection` the subquery outputs for coercion? Like would this need to be an `Collection.Array` or `Collection.Bag` is also allowed?

Reply:
Ah, `Rex.Collection` is different than `Rex.Query.Collection`.

`Rex.Collection` is a literal `Array` or `Bag`. `Rex.Query.Collection` is a `SFW` (or `SVFW`) returning a collection.